File: cpp/src/qmc.cpp
// qmc.cpp -- implementation of qmc.hpp.

#include "qmc.hpp"

#include "black_scholes.hpp"   // norm_cdf if needed (not used directly here)
#include "gbm.hpp"              // inverse_normal_cdf, validators

#include <algorithm>
#include <cmath>
#include <cstring>
#include <fstream>
#include <limits>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace quant {

namespace {

// First 20 prime numbers, used as Halton bases.
const int FIRST_PRIMES[20] = {
    2,  3,  5,  7, 11, 13, 17, 19, 23, 29,
   31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
};

// Compute the base-b radical inverse of i.
// Iterative implementation; bounded floating-point accumulator.
double radical_inverse(std::uint64_t i, int base) {
    double result = 0.0;
    double f = 1.0 / static_cast<double>(base);
    while (i > 0) {
        result += f * static_cast<double>(i % static_cast<std::uint64_t>(base));
        i /= static_cast<std::uint64_t>(base);
        f /= static_cast<double>(base);
    }
    return result;
}

// Find the rightmost zero bit of i (1-based index).
// Used for the Sobol Gray-code recursion: u_{i+1} = u_i XOR V[c],
// where c is the position of the rightmost zero bit of i.
//
// For i = 0 (binary ...0), rightmost zero bit is position 1.
// For i = 1 (binary ...01), rightmost zero bit is position 2.
// For i = 2 (binary ...10), rightmost zero bit is position 1.
// For i = 3 (binary ...11), rightmost zero bit is position 3.
// Etc.
inline int rightmost_zero_bit(std::uint64_t i) {
    // Equivalent to: position of lowest 1 bit of (~i), 1-based.
    // For Sobol Gray code, this returns 1 + ctz(~i).
    int pos = 1;
    while (i & 1u) {
        i >>= 1;
        ++pos;
    }
    return pos;
}

// Convert a Sobol-state integer to a double in [0, 1).
// State integers are stored left-shifted to occupy the high bits;
// dividing by 2^64 (as a floating value) gives a value in [0, 1).
inline double state_to_double(std::uint64_t x) {
    // 2^64 as a double (exact since 2^64 is representable in float64).
    constexpr double TWO_64 = 18446744073709551616.0;
    return static_cast<double>(x) / TWO_64;
}

}  // anonymous namespace


// =====================================================================
// Halton implementation
// =====================================================================

Halton::Halton(std::size_t d) : d_(d), index_(1) {
    if (d == 0) {
        throw std::invalid_argument("Halton: d must be >= 1");
    }
    if (d > 20) {
        throw std::invalid_argument(
            "Halton: d > 20 not supported (would need primes beyond 71)");
    }
    primes_.assign(FIRST_PRIMES, FIRST_PRIMES + d);
}

void Halton::generate(std::size_t n, double* out) {
    for (std::size_t i = 0; i < n; ++i) {
        for (std::size_t j = 0; j < d_; ++j) {
            out[i * d_ + j] = radical_inverse(index_ + i, primes_[j]);
        }
    }
    index_ += n;
}


// =====================================================================
// Sobol implementation
// =====================================================================

Sobol::Sobol(std::size_t d, const std::string& data_file)
    : d_(d), index_(1) {
    if (d == 0) {
        throw std::invalid_argument("Sobol: d must be >= 1");
    }
    if (d > MAX_DIM) {
        throw std::invalid_argument(
            "Sobol: d > " + std::to_string(MAX_DIM) +
            " not supported in this build");
    }

    // Number of bits in the Sobol state. We use 32: enough for
    // 2^32 ~ 4 billion points, more than ample for our budgets.
    constexpr int BITS = 32;

    V_.assign(d_, std::vector<std::uint64_t>(BITS + 1, 0));

    // Dimension 1 is special: m_k = 1 for all k, V_k = 2^(BITS - k).
    // (This corresponds to the trivial polynomial "x + 0", i.e. no
    // primitive polynomial information needed.)
    for (int k = 1; k <= BITS; ++k) {
        V_[0][k] = static_cast<std::uint64_t>(1) << (BITS - k);
        // Shift to high bits of uint64_t for state storage.
        V_[0][k] <<= (64 - BITS);
    }

    if (d_ > 1) {
        load_direction_numbers(data_file);
    }

    x_.assign(d_, 0);
}

void Sobol::load_direction_numbers(const std::string& data_file) {
    std::ifstream in(data_file);
    if (!in) {
        throw std::runtime_error(
            "Sobol: cannot open direction-numbers file: " + data_file);
    }

    constexpr int BITS = 32;

    // Skip the header line.
    std::string header;
    std::getline(in, header);

    // Read dimensions 2, 3, ..., d_.
    for (std::size_t j = 1; j < d_; ++j) {
        std::string line;
        if (!std::getline(in, line)) {
            throw std::runtime_error(
                "Sobol: direction-numbers file ended at dimension " +
                std::to_string(j + 1));
        }
        std::istringstream iss(line);

        std::size_t d_index;
        int s;          // degree of primitive polynomial
        std::uint64_t a;  // polynomial coefficient encoding
        iss >> d_index >> s >> a;
        if (!iss) {
            throw std::runtime_error(
                "Sobol: malformed line for dimension " +
                std::to_string(j + 1));
        }

        // Read the s initial direction numbers m_1, ..., m_s.
        std::vector<std::uint64_t> m(s + 1, 0);
        for (int k = 1; k <= s; ++k) {
            iss >> m[k];
        }

        // Extend the m sequence using the recurrence:
        //   m_k = 2 a_1 m_{k-1} XOR 2^2 a_2 m_{k-2} XOR ... XOR
        //         2^{s-1} a_{s-1} m_{k-s+1} XOR (2^s m_{k-s} XOR m_{k-s})
        // for k > s.
        //
        // The polynomial coefficients a_1, ..., a_{s-1} are encoded
        // in `a` (s-1 bits, with a_1 being the MSB).
        for (int k = s + 1; k <= BITS; ++k) {
            std::uint64_t m_k = m[k - s] ^ (m[k - s] << s);
            for (int i = 1; i <= s - 1; ++i) {
                // Extract a_i from the encoding `a`.
                int a_i = static_cast<int>((a >> (s - 1 - i)) & 1);
                if (a_i) {
                    m_k ^= m[k - i] << i;
                }
            }
            m.push_back(m_k);
        }

        // Convert m_k to direction numbers V_k by shifting:
        //   V_k = m_k * 2^{BITS - k}
        // and store left-shifted into the high bits of uint64_t.
        for (int k = 1; k <= BITS; ++k) {
            V_[j][k] = m[k] << (BITS - k);
            V_[j][k] <<= (64 - BITS);
        }
    }
}

void Sobol::reset() {
    index_ = 1;
    std::fill(x_.begin(), x_.end(), 0);
}

void Sobol::generate(std::size_t n, double* out) {
    // We use the Gray-code recursion. State x_[j] holds the current
    // Sobol point in dimension j (as a high-bit-shifted uint64_t).
    // To advance from index i to index i+1: x XOR V_[j][c], where c
    // is the rightmost-zero-bit position of i.
    //
    // Index 0 (the all-zero point) is degenerate under Phi^{-1}, so
    // we skip it: index_ starts at 1 in the constructor, meaning the
    // first call to generate produces points 1, 2, 3, ...
    //
    // Building the i-th point from scratch via XOR over the binary
    // expansion of i requires up to log2(i) operations. The
    // Gray-code shortcut needs only one XOR per dimension per step.
    // To get to index_ = N from a fresh state, we need to advance
    // sequentially from 0; reset() does this implicitly.

    for (std::size_t i = 0; i < n; ++i) {
        // Update state from index_-1 to index_. Use Gray code: c is
        // the position of the rightmost zero bit of (index_ - 1).
        std::uint64_t prev = index_ - 1;
        int c = rightmost_zero_bit(prev);

        for (std::size_t j = 0; j < d_; ++j) {
            x_[j] ^= V_[j][c];
            out[i * d_ + j] = state_to_double(x_[j]);
        }
        ++index_;
    }
}


// =====================================================================
// Digital shift
// =====================================================================

void digital_shift(std::size_t n, std::size_t d,
                   const double* shift,
                   double* points) {
    // Convert to 53-bit unsigned integers, XOR, convert back.
    constexpr double SCALE = 9007199254740992.0;  // 2^53
    constexpr double INV_SCALE = 1.0 / SCALE;

    // Pre-convert shift.
    std::vector<std::uint64_t> shift_int(d);
    for (std::size_t j = 0; j < d; ++j) {
        shift_int[j] = static_cast<std::uint64_t>(shift[j] * SCALE);
    }

    for (std::size_t i = 0; i < n; ++i) {
        for (std::size_t j = 0; j < d; ++j) {
            const std::uint64_t u_int = static_cast<std::uint64_t>(
                points[i * d + j] * SCALE);
            const std::uint64_t shifted = u_int ^ shift_int[j];
            points[i * d + j] = static_cast<double>(shifted) * INV_SCALE;
        }
    }
}


// =====================================================================
// Helper: convert (n_paths, n_steps) uniforms into discounted Euler
// payoffs for the European call.
// =====================================================================

namespace {

// Generates n_paths discounted European-call payoffs from an n_paths-by-n_steps
// row-major matrix of uniform points u in [0, 1)^{n_steps}. Inverts each u_{i,k}
// to a standard normal Z via inverse_normal_cdf, runs the Euler step over
// k = 0, ..., n_steps - 1, and returns the discounted payoff per path.
//
// Clipping: avoid u in {0, 1} which would map to +/- infinity under Phi^{-1}.
void euler_payoffs(const double* u, std::size_t n_paths, std::size_t n_steps,
                   double S, double K, double r, double sigma, double T,
                   std::vector<double>& out) {
    const double h = T / static_cast<double>(n_steps);
    const double sqrt_h = std::sqrt(h);
    const double discount = std::exp(-r * T);
    const double EPS = 1e-15;

    out.resize(n_paths);

    for (std::size_t i = 0; i < n_paths; ++i) {
        double S_path = S;
        for (std::size_t k = 0; k < n_steps; ++k) {
            double u_ik = u[i * n_steps + k];
            // Clip away from 0 and 1.
            if (u_ik < EPS) u_ik = EPS;
            if (u_ik > 1.0 - EPS) u_ik = 1.0 - EPS;

            const double Z = inverse_normal_cdf(u_ik);
            S_path *= 1.0 + r * h + sigma * sqrt_h * Z;
        }
        out[i] = discount * std::max(S_path - K, 0.0);
    }
}

}  // anonymous namespace


// =====================================================================
// Deterministic QMC pricer
// =====================================================================

double mc_european_call_euler_qmc(double S, double K, double r,
                                  double sigma, double T,
                                  std::size_t n_paths,
                                  std::size_t n_steps,
                                  const std::string& sequence) {
    validate_model_params(S, sigma, T);
    validate_strike(K);
    validate_n_paths(n_paths);
    validate_n_steps(n_steps);

    std::vector<double> u(n_paths * n_steps);

    if (sequence == "halton") {
        Halton h(n_steps);
        h.generate(n_paths, u.data());
    } else if (sequence == "sobol") {
        Sobol s(n_steps);
        s.generate(n_paths, u.data());
    } else {
        throw std::invalid_argument(
            "mc_european_call_euler_qmc: sequence must be 'halton' or "
            "'sobol', got '" + sequence + "'");
    }

    std::vector<double> payoffs;
    euler_payoffs(u.data(), n_paths, n_steps, S, K, r, sigma, T, payoffs);

    double sum = 0.0;
    for (double p : payoffs) sum += p;
    return sum / static_cast<double>(n_paths);
}


// =====================================================================
// RQMC pricer with digital shift
// =====================================================================

MCResult
mc_european_call_euler_rqmc(double S, double K, double r,
                            double sigma, double T,
                            std::size_t n_paths,
                            std::size_t n_steps,
                            std::size_t n_replications,
                            std::mt19937_64& rng,
                            double confidence_level) {
    validate_model_params(S, sigma, T);
    validate_strike(K);
    validate_n_paths(n_paths);
    validate_n_steps(n_steps);
    validate_confidence_level(confidence_level);

    if (n_replications < 2) {
        throw std::invalid_argument(
            "mc_european_call_euler_rqmc: n_replications must be >= 2 "
            "for a meaningful sample variance");
    }

    // Generate the base Sobol point set ONCE.
    Sobol sob(n_steps);
    std::vector<double> u_base(n_paths * n_steps);
    sob.generate(n_paths, u_base.data());

    // Build R replication estimates.
    std::uniform_real_distribution<double> uniform(0.0, 1.0);
    std::vector<double> replication_estimates(n_replications);
    std::vector<double> u_shifted(n_paths * n_steps);
    std::vector<double> shift(n_steps);
    std::vector<double> payoffs;

    for (std::size_t r_idx = 0; r_idx < n_replications; ++r_idx) {
        // Draw a fresh shift.
        for (std::size_t j = 0; j < n_steps; ++j) {
            shift[j] = uniform(rng);
        }

        // Copy and shift.
        std::memcpy(u_shifted.data(), u_base.data(),
                    sizeof(double) * u_base.size());
        digital_shift(n_paths, n_steps, shift.data(), u_shifted.data());

        // Compute Euler payoffs.
        euler_payoffs(u_shifted.data(), n_paths, n_steps,
                      S, K, r, sigma, T, payoffs);

        // Mean of payoffs = this replication's estimate.
        double sum = 0.0;
        for (double p : payoffs) sum += p;
        replication_estimates[r_idx] = sum / static_cast<double>(n_paths);
    }

    // Reduce: estimate, sample variance, half-width over the R
    // replications. The CLT applies to these R i.i.d. estimates.
    return mc_estimator(replication_estimates, confidence_level);
}

}  // namespace quant

File: python/quantlib/qmc.py
"""Quasi-Monte Carlo and Randomized QMC pricers for the European call.

This module implements the Block 3 part of the variance-reduction
toolkit: deterministic QMC sequences (Halton, Sobol) for pricing the
European call under the Euler discretisation, plus a randomized
QMC variant with digital shifting that recovers a usable half-width.

The integration problem is d-dimensional with d = N (the number of
Euler steps). We use d = 20 as the canonical setup for validation;
this is dimensional enough to exhibit non-trivial QMC behaviour
without entering the high-dimensional regime where QMC's logarithmic
factor catastrophe dominates.

Functions
---------
halton(n, d):
    First n entries of the d-dimensional Halton sequence, from
    scratch via radical inverse. Returns shape (n, d) array.

sobol(n, d, scramble=False, seed=None):
    First n entries of the d-dimensional Sobol sequence, via
    scipy.stats.qmc.Sobol (which uses Joe-Kuo tables internally).
    If scramble=True, returns a digitally shifted version.

mc_european_call_euler_qmc:
    Deterministic QMC pricer using either Halton or Sobol. Returns
    a single estimate; no half-width because no statistical model.

mc_european_call_euler_rqmc:
    Randomized QMC pricer using Sobol with R independent digital
    shifts. Returns a full MCResult (n_paths field stores R, the
    number of replications, not the n*R individual payoffs).

References
----------
Phase 2 Block 3.0 writeup (foundations); Block 3.1 writeup (this
implementation). Glasserman, *Monte Carlo Methods in Financial
Engineering*, Chapter 5.
"""

import math

import numpy as np
from scipy.stats import qmc, norm

from quantlib.gbm import (
    validate_model_params,
    validate_strike,
    validate_n_paths,
    validate_n_steps,
)
from quantlib.monte_carlo import MCResult


# =====================================================================
# Halton sequence (from scratch)
# =====================================================================

# The first 20 prime numbers, used as the bases for the d-dimensional
# Halton sequence. Beyond d = 20 we would need to extend this list,
# but the construction would degrade rapidly anyway: for d > 8 Halton
# is known to exhibit visible projection artefacts.
_FIRST_PRIMES = [
    2, 3, 5, 7, 11, 13, 17, 19, 23, 29,
    31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
]


def _radical_inverse(i: int, base: int) -> float:
    """The base-b radical inverse of the integer i.

    Writes i in base b, reverses the digits, places a decimal point.
    Returns a value in [0, 1).

    Computed iteratively to avoid forming large powers and the
    associated floating-point loss; the running accumulator stays
    bounded by 1.
    """
    result = 0.0
    f = 1.0 / base
    while i > 0:
        result += f * (i % base)
        i //= base
        f /= base
    return result


def halton(n: int, d: int) -> np.ndarray:
    """Generate the first n entries of the d-dimensional Halton sequence.

    Each coordinate j uses the prime base p_{j+1} (so coordinate 0
    uses base 2, coordinate 1 uses base 3, etc.). The first index
    skipped is 0 (which would give the all-zero point); we start at
    i = 1 to avoid this degenerate point.

    Returns
    -------
    np.ndarray of shape (n, d), entries in [0, 1).
    """
    if d > len(_FIRST_PRIMES):
        raise ValueError(
            f"Halton sequence supports at most d = {len(_FIRST_PRIMES)} "
            f"dimensions (would need primes beyond {_FIRST_PRIMES[-1]})"
        )
    if n < 1:
        raise ValueError("n must be at least 1")

    out = np.empty((n, d), dtype=np.float64)
    for j in range(d):
        base = _FIRST_PRIMES[j]
        # Start at index 1: index 0 produces the all-zero point.
        for i in range(n):
            out[i, j] = _radical_inverse(i + 1, base)
    return out


# =====================================================================
# Sobol sequence (wrapper around scipy.stats.qmc.Sobol)
# =====================================================================

def sobol(n: int, d: int, *, scramble: bool = False,
          seed=None) -> np.ndarray:
    """Generate the first n entries of the d-dimensional Sobol sequence.

    Uses scipy.stats.qmc.Sobol, which loads the Joe-Kuo direction-number
    tables internally (covering dimensions up to 21201).

    Parameters
    ----------
    n : int
        Number of points. Sobol's optimal points are at powers of 2;
        for n = 2^k the equidistribution is best. For non-power-of-2
        n, scipy returns the first n points without warning, but the
        practitioner should prefer powers of 2 when possible.
    d : int
        Dimension.
    scramble : bool, default False
        If True, applies Owen scrambling (a randomisation more
        sophisticated than digital shift). For our deterministic-QMC
        pricer we use scramble=False; for RQMC we apply our own
        digital shift externally.
    seed : optional
        Seed for the scrambling, if scramble=True.

    Returns
    -------
    np.ndarray of shape (n, d), entries in [0, 1).
    """
    if n < 1:
        raise ValueError("n must be at least 1")
    if d < 1:
        raise ValueError("d must be at least 1")

    sampler = qmc.Sobol(d=d, scramble=scramble, seed=seed)
    return sampler.random(n)


# =====================================================================
# Deterministic QMC pricer
# =====================================================================

def _euler_payoffs(u: np.ndarray, S: float, K: float, r: float,
                   sigma: float, T: float) -> np.ndarray:
    """Convert a (n, N) array of uniform points into n discounted Euler
    European-call payoffs.

    Each row of u becomes one path: u[i, k] -> Z_k = Phi^{-1}(u[i, k]),
    then S_T^{(N)} accumulates via the Euler recursion.
    """
    n, N = u.shape
    h = T / N
    discount = math.exp(-r * T)

    # Convert uniforms to standard normals via inversion.
    # ppf is the inverse normal CDF (Phi^{-1}).
    Z = norm.ppf(u)

    # Euler recursion: S_{k+1} = S_k * (1 + r*h + sigma*sqrt(h)*Z_k).
    # Vectorised over the n paths simultaneously.
    sqrt_h = math.sqrt(h)
    S_path = np.full(n, S, dtype=np.float64)
    for k in range(N):
        S_path *= 1.0 + r * h + sigma * sqrt_h * Z[:, k]

    return discount * np.maximum(S_path - K, 0.0)


def mc_european_call_euler_qmc(S, K, r, sigma, T, n_paths,
                               *,
                               n_steps: int = 20,
                               sequence: str = "sobol"):
    """Deterministic QMC pricer for the European call (Euler).

    Generates a single QMC point set in dimension d = n_steps,
    inverts to normals via Phi^{-1}, runs the Euler scheme, returns
    the average discounted payoff. No half-width: the estimator is
    deterministic.

    Parameters
    ----------
    sequence : 'halton' or 'sobol'
        Which low-discrepancy sequence to use.

    Returns
    -------
    float
        The deterministic QMC estimate of the European call price.

    Notes
    -----
    Returning a bare float (not an MCResult) is deliberate: an
    MCResult with a half-width of zero would be misleading, since
    the estimator has unknown statistical error (the Koksma-Hlawka
    bound requires V(f), which is intractable for option payoffs).
    The user is forced to acknowledge the lack of an error bar.
    For a proper error estimate, use mc_european_call_euler_rqmc.
    """
    validate_model_params(S, r, sigma, T)
    validate_strike(K)
    validate_n_paths(n_paths)
    validate_n_steps(n_steps)

    if sequence == "halton":
        u = halton(n_paths, n_steps)
    elif sequence == "sobol":
        u = sobol(n_paths, n_steps, scramble=False)
    else:
        raise ValueError(
            f"sequence must be 'halton' or 'sobol', got {sequence!r}"
        )

    # Avoid u = 0 or u = 1, which would map to +/- infinity under
    # Phi^{-1}. This can happen at the first Sobol point (which is
    # the origin) or at the radical inverse boundaries. We clip into
    # the open interval with a small epsilon.
    eps = np.finfo(np.float64).tiny
    u = np.clip(u, eps, 1.0 - eps)

    payoffs = _euler_payoffs(u, S, K, r, sigma, T)
    return float(np.mean(payoffs))


# =====================================================================
# RQMC pricer with digital shift
# =====================================================================

def _digital_shift(u: np.ndarray, shift: np.ndarray) -> np.ndarray:
    """Apply a digital shift (XOR on the binary expansions) to a QMC
    point set.

    For each entry u_{i,j} and shift component xi_j, both in [0, 1),
    convert to 53-bit integers (the mantissa precision of float64),
    XOR, and convert back. This is the standard digital shift in
    floating-point form.

    Parameters
    ----------
    u : ndarray of shape (n, d), values in [0, 1).
    shift : ndarray of shape (d,), values in [0, 1).

    Returns
    -------
    ndarray of shape (n, d) with each row digitally shifted by
    `shift`.
    """
    # Convert to 53-bit unsigned integers via bit-shifting.
    # A float in [0, 1) maps to an integer in [0, 2^53).
    SCALE = float(1 << 53)
    u_int = (u * SCALE).astype(np.uint64)
    shift_int = (shift * SCALE).astype(np.uint64)

    # XOR (broadcast: shift_int has shape (d,), u_int has shape (n, d)).
    shifted_int = u_int ^ shift_int[np.newaxis, :]

    # Convert back to float in [0, 1).
    return shifted_int.astype(np.float64) / SCALE


def mc_european_call_euler_rqmc(S, K, r, sigma, T, n_paths,
                                *,
                                n_steps: int = 20,
                                n_replications: int = 20,
                                seed=None,
                                confidence_level: float = 0.95):
    """Randomized QMC pricer for the European call (Euler) with
    Sobol + digital shift.

    Procedure:
      1. Generate a single deterministic Sobol point set of size n.
      2. For each of R replications:
         a. Draw a uniform random shift xi in [0, 1)^d.
         b. XOR the Sobol points with xi to get a shifted point set.
         c. Run the Euler pricer on the shifted points.
      3. The R replication estimates are i.i.d.; their sample mean
         and sample variance give the RQMC estimate and half-width.

    Parameters
    ----------
    n_paths : int
        Per-replication number of QMC points. The total payoff budget
        is n_paths * n_replications.
    n_replications : int, default 20
        Number of independent digital shifts.
    seed : optional
        Seed for the shifts. The Sobol point set itself is deterministic
        regardless of the seed.

    Returns
    -------
    MCResult
        result.estimate is the average of R replication estimates.
        result.half_width is z * sigma_rep / sqrt(R), where sigma_rep
        is the empirical sample sd across the R replications.
        result.n_paths stores R (the number of i.i.d. units), not
        n_paths * n_replications.
    """
    validate_model_params(S, r, sigma, T)
    validate_strike(K)
    validate_n_paths(n_paths)
    validate_n_steps(n_steps)
    if n_replications < 2:
        raise ValueError("n_replications must be at least 2 for a "
                         "meaningful sample variance")

    rng = np.random.default_rng(seed)

    # Generate the base Sobol point set ONCE, deterministically.
    u_base = sobol(n_paths, n_steps, scramble=False)
    eps = np.finfo(np.float64).tiny
    u_base = np.clip(u_base, eps, 1.0 - eps)

    # Build R replication estimates.
    replication_estimates = np.empty(n_replications, dtype=np.float64)
    for r_idx in range(n_replications):
        shift = rng.random(n_steps)
        u_shifted = _digital_shift(u_base, shift)
        # Re-clip after XOR: in rare cases the XOR can produce exactly 0.
        u_shifted = np.clip(u_shifted, eps, 1.0 - eps)

        payoffs = _euler_payoffs(u_shifted, S, K, r, sigma, T)
        replication_estimates[r_idx] = float(np.mean(payoffs))

    # Reduce: estimate, sample variance, half-width over the R
    # replications. The CLT applies to these R i.i.d. estimates.
    estimate = float(np.mean(replication_estimates))
    sample_variance = float(np.var(replication_estimates, ddof=1))
    z = float(norm.ppf(0.5 + confidence_level / 2.0))
    half_width = z * math.sqrt(sample_variance / n_replications)

    return MCResult(
        estimate=estimate,
        half_width=half_width,
        sample_variance=sample_variance,
        n_paths=n_replications,
    )


# =====================================================================
# Smoke test entry point
# =====================================================================

if __name__ == "__main__":
    S, K, r, sigma, T = 100.0, 100.0, 0.05, 0.20, 1.00
    N = 20

    # Deterministic QMC pricers. Using a power of 2 for Sobol's
    # optimal balance properties (no warning emitted by scipy).
    est_halton = mc_european_call_euler_qmc(
        S, K, r, sigma, T, n_paths=8192,
        n_steps=N, sequence="halton",
    )
    print(f"QMC Halton (n=8192, N={N}): estimate = {est_halton:.6f}")

    est_sobol = mc_european_call_euler_qmc(
        S, K, r, sigma, T, n_paths=8192,
        n_steps=N, sequence="sobol",
    )
    print(f"QMC Sobol  (n=8192, N={N}): estimate = {est_sobol:.6f}")

    # RQMC pricer.
    result_rqmc = mc_european_call_euler_rqmc(
        S, K, r, sigma, T, n_paths=4096,
        n_steps=N, n_replications=20, seed=42,
    )
    print(f"\nRQMC Sobol (n=4096, R=20, N={N}, total payoffs=81920):")
    print(f"  estimate     : {result_rqmc.estimate:.6f}")
    print(f"  half-width   : {result_rqmc.half_width:.6f}")
    print(f"  sample var   : {result_rqmc.sample_variance:.6f}")
    print(f"  n (= R)      : {result_rqmc.n_paths}")
